### PR TITLE
Towers of Hanoi Eval

### PR DESCRIPTION
**Pull request by bhdoggett**

### CODE REVIEW

**Comment by bhdoggett** [comment]
hanoi.js contains the basic code with the standard hanoi board with 5 discs and 3 pegs. main.js contains code which enables the player to initialize different boards (and name them differently) with chosen numbers of discs and pegs.

File: index.html [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<!doctype html>
+<html lang="en">
+  <head>
+    <meta charset="UTF-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Towers of Hanoi</title>
+    <script src="towers.js"></script>
```
Comment (ronyrosenberg) [inline index.html]: wrong file name. There is no towers.js file

File: hanoi.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
+// console.log("let's do this!");
+
+let towers = [];
+
+let initializeBoard = (pegs, discs) => {
```
Comment (ronyrosenberg) [inline hanoi.js]: should be a const

File: hanoi.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
+// console.log("let's do this!");
+
+let towers = [];
```
Comment (ronyrosenberg) [inline hanoi.js]: should be const if its not being reassigned

File: hanoi.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+let initializeBoard = (pegs, discs) => {
+  pegs = Number(prompt("How many pegs?"));
+  discs = Number(prompt("How many discs?"));
```
Comment (ronyrosenberg) [inline hanoi.js]: VERY cool idea, but add some basic validation like 3 pegs minimum

File: hanoi.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+  for (let i = 0; i < pegs; i++) {
```
Comment (ronyrosenberg) [inline hanoi.js]: DRY, very important principle. Extract to a function and reuse instead of repeating the code

File: hanoi.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  console.log(
+    `LET THE GAME BEGIN! Your board is set with ${pegs} pegs and ${discs} discs:`
+  );
+  hanoi.printBoard();
+};
+
+const hanoi = {
+  board: { 1: towers[0], 2: towers[1], 3: towers[2] },
+  printBoard: () => {
+    towers.map((tower) => console.log(tower));
```
Comment (ronyrosenberg) [inline hanoi.js]: wrong usage of map, use forEach as you are not storing what map is returning.

File: hanoi.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+  },
+};
+
+const moveDisc = (from, to) => {
+  let possible = true;
```
Comment (ronyrosenberg) [inline hanoi.js]: good use of let in this case

File: hanoi.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    towers[from - 1].splice(towers[from - 1].length - 1, 1);
+    console.log("That move was successful, board is now:");
+    hanoi.printBoard();
+  } else {
+    console.log(
+      "You cannont move a larger disc on top of a smaller one, the board is still:"
+    );
+    hanoi.printBoard();
+
+    if (to !== 1 && towers[to - 1].length === 5) {
```
Comment (ronyrosenberg) [inline hanoi.js]: this is a bug, what if I set 3 pegs? I didn't win! but I should have.
image

File: hanoi.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+  // Calcule the correct tower to move the disk FROM with: towers[from - 1];
+  // Calculate the size of the disk in the FROM tower with: towers[from - 1][towers[from - 1].length - 1];
+
+  // Calcule the correct tower to move the disk TO with: towers[to - 1];
+  // Calculate the size of the disk in the TO tower with: towers[to - 1][towers[to - 1].length - 1];
+
+  if (
+    towers[from - 1][towers[from - 1].length - 1] >
+    towers[to - 1][towers[to - 1].length - 1]
```
Comment (ronyrosenberg) [inline hanoi.js]: not enough conditions. You are missing, moving from a peg that has no discs, and moving from a peg to the same peg.

File: hanoi.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  if (possible === true) {
+    towers[to - 1].push(towers[from - 1][towers[from - 1].length - 1]);
+    towers[from - 1].splice(towers[from - 1].length - 1, 1);
+    console.log("That move was successful, board is now:");
+    hanoi.printBoard();
+  } else {
+    console.log(
+      "You cannont move a larger disc on top of a smaller one, the board is still:"
```
Comment (ronyrosenberg) [inline hanoi.js]: your error messages should be more complex to attend the different use cases, classic example for a dictionary
```suggestion
{
'errorKey1': "You cannont move a larger disc on top of a smaller one, the board is still",
'errorKey2': "You cannot move a disc to the same original peg",
etc...

}
```

File: main.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
+// console.log("let's do this!");
+
+const Towers = function (name, pegs, discs) {
```
Comment (ronyrosenberg) [inline main.js:1]: cool implementation, You can review it with your mentor 💯 🥇 
If you still want it rereviewed and reevaluated, submit it again with only this file. You get 3 attempts per eval

File: hanoi.js [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  console.log(
+    `LET THE GAME BEGIN! Your board is set with ${pegs} pegs and ${discs} discs:`
+  );
+  hanoi.printBoard();
+};
+
+const hanoi = {
+  board: { 1: towers[0], 2: towers[1], 3: towers[2] },
```
Comment (ronyrosenberg) [inline hanoi.js]: what is this? remove if not in use, its redundant.

File: hanoi.js (hunk 9 above)
Comment (ronyrosenberg) [reply hanoi.js]: Also should be moved to a function

File: hanoi.js [hunk 13]
```diff
@@ -0,0 +1,66 @@
+// console.log("let's do this!");
```
Comment (ronyrosenberg) [inline hanoi.js]: do not commit commented code

File: index.html (hunk 1 above)
Comment (bhdoggett) [reply index.html]: Thanks, updated!

File: hanoi.js (hunk 2 above)
Comment (bhdoggett) [reply hanoi.js]: Thanks I checked my variables in main.js as well to make sure  I'm only using let when I explicitly want the variable's value to be able to change.

File: main.js [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
```
Comment (ronyrosenberg) [inline main.js:1]: Would have used a class approach with a constructor to define this.

File: main.js [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+    for (let i = 0; i < this.pegs; i++) {
+      towers.push([]);
+    }
+
+    for (let j = this.discs; j > 0; j--) {
+      towers[0].push(j);
+    }
+
+    return towers;
+  };
```
Comment (ronyrosenberg) [inline main.js:18]: this function can be outside and be invoked from here

File: main.js [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+  this.towers = this.initializeBoard();
+  this.printBoard = function () {
```
Comment (ronyrosenberg) [inline main.js:20]: why not es6? arrow function

Comment (ronyrosenberg) [reply main.js:20]: also better to define outside in case you need to reuse

File: main.js [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+    console.log(message.error2);
+  }
+
+  if (possible) {
+    this.towers[to - 1].push(
+      this.towers[from - 1][this.towers[from - 1].length - 1]
+    );
+    this.towers[from - 1].splice(this.towers[from - 1].length - 1, 1);
+
+    if (to !== 1 && this.towers[to - 1].length === this.discs) {
```
Comment (ronyrosenberg) [inline main.js:56]: its a less sustainable approach to nest conditions, makes the code hard to follow.

File: main.js [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+      console.log(message.successfulMove);
+    }
+  }
+
+  this.moveCount++;
+  this.printBoard();
+};
+
+Towers.prototype.autoPlay = function (numMoves) {
+  const randomMove = () => {
```
Comment (ronyrosenberg) [inline main.js:71]: another bad use of scope, better to define outside.

File: main.js [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+    const moveFrom = Math.ceil(Math.random() * this.towers.length);
+    const moveTo = Math.ceil(Math.random() * this.towers.length);
```
Comment (ronyrosenberg) [inline main.js:73]: DRY

File: main.js [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+    } else {
+      console.log(message.successfulMove);
+    }
+  }
+
+  this.moveCount++;
+  this.printBoard();
+};
+
+Towers.prototype.autoPlay = function (numMoves) {
```
Comment (ronyrosenberg) [inline main.js:70]: arrow function

File: main.js [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+    const moveFrom = Math.ceil(Math.random() * this.towers.length);
+    const moveTo = Math.ceil(Math.random() * this.towers.length);
+
+    play = this.moveDisc(moveFrom, moveTo);
+  };
+
+  for (let i = 0; i < numMoves; i++) {
+    randomMove();
+    this.moveCount++;
+    if (!this.shouldAutoPlay) {
```
Comment (ronyrosenberg) [inline main.js:81]: classic approach to use while(condition).... do something instead of for loop
